--- courier-imap-5.2.2.org/configure.ac	2019-03-27 22:34:04.000000000 +0100
+++ courier-imap-5.2.2/configure.ac	2021-10-23 00:10:21.114488946 +0200
@@ -268,6 +268,7 @@ AC_CONFIG_FILES(Makefile dbobj.h dbobj.c
 		courier-imap.spec
 		courier-debuild
 		INSTALL.html
+		courier-imap courier-imap-ssl courier-pop3 courier-pop3-ssl
 		packaging/suse/courier-imap.init
 		packaging/systemd/courier-imap.sysvinit
 		packaging/systemd/courier-imap.service)
