--- courier-imap-4.4.1.orig/Makefile.am	2008-07-15 00:26:58.000000000 +0200
+++ courier-imap-4.4.1/Makefile.am	2008-10-08 08:23:14.968691362 +0200
@@ -10,7 +10,7 @@
 	packaging/suse/pop3.pam
 
 SUBDIRS=numlib md5 sha1 bdbobj gdbmobj soxwrap unicode rfc822 random128 \
-	rfc1035 rfc2045 liblock maildir makedat libhmac \
+	rfc1035 rfc2045 liblock maildir libhmac \
 	waitlib tcpd rootcerts imap
 
 sbin_SCRIPTS=sharedindexinstall sharedindexsplit
@@ -20,7 +20,7 @@
 
 binPROGRAMS=imapd pop3d maildirmake maildiracl deliverquota maildirkw
 sbinPROGRAMS=imaplogin pop3login
-libexecPROGRAMS=makedatprog couriertcpd
+libexecPROGRAMS=couriertcpd
 
 bin_PROGRAMS= @binPROGRAMS_exec@
 sbin_PROGRAMS=@sbinPROGRAMS_exec@
